src/processor/operator/recursive_extend/shortest_path_recursive_join.cpp

```diff
@@ -0,0 +1,69 @@
+#include "processor/operator/recursive_extend/shortest_path_recursive_join.h"
+
+namespace kuzu {
+namespace processor {
+
+void ShortestPathRecursiveJoin::initLocalStateInternal(
+    ResultSet* resultSet_, ExecutionContext* context) {
+    BaseRecursiveJoin::initLocalStateInternal(resultSet_, context);
+    distanceVector = resultSet->getValueVector(distanceVectorPos);
+    auto maxNodeOffset = nodeTable->getMaxNodeOffset(transaction);
+    bfsMorsel = std::make_unique<ShortestPathBFSMorsel>(
+        maxNodeOffset, lowerBound, upperBound, sharedState->semiMask.get());
+    bfsMorsel->resetState();
+}
+
+bool ShortestPathRecursiveJoin::scanOutput() {
+    auto morsel = (ShortestPathBFSMorsel*)bfsMorsel.get();
+    if (outputCursor == morsel->dstNodeOffsets.size()) {
+        return false;
+    }
+    auto vectorSize = 0u;
+    while (vectorSize != common::DEFAULT_VECTOR_CAPACITY &&
+           outputCursor < morsel->dstNodeOffsets.size()) {
+        auto offset = morsel->dstNodeOffsets[outputCursor];
+        dstNodeIDVector->setValue<common::nodeID_t>(
+            vectorSize, common::nodeID_t{offset, nodeTable->getTableID()});
+        distanceVector->setValue<int64_t>(vectorSize, morsel->dstNodeOffset2PathLength.at(offset));
+        vectorSize++;
+        outputCursor++;
+    }
+    dstNodeIDVector->state->initOriginalAndSelectedSize(vectorSize);
+    return true;
+}
+
+void ShortestPathRecursiveJoin::computeBFS(kuzu::processor::ExecutionContext* context) {
+    auto morsel = (ShortestPathBFSMorsel*)bfsMorsel.get();
+    auto nodeID = srcNodeIDVector->getValue<common::nodeID_t>(
+        srcNodeIDVector->state->selVector->selectedPositions[0]);
+    morsel->markSrc(nodeID.offset);
+    while (!morsel->isComplete()) {
+        auto nodeOffset = morsel->getNextNodeOffset();
+        if (nodeOffset != common::INVALID_NODE_OFFSET) {
+            // Found a starting node from current frontier.
+            scanBFSLevel->setNodeID(common::nodeID_t{nodeOffset, nodeTable->getTableID()});
+            while (recursiveRoot->getNextTuple(context)) { // Exhaust recursive plan.
+                updateVisitedState(morsel);
+            }
+        } else {
+            // Otherwise move to the next frontier.
+            morsel->moveNextLevelAsCurrentLevel();
+        }
+    }
+}
+
+void ShortestPathRecursiveJoin::updateVisitedState(ShortestPathBFSMorsel* morsel) {
```

Comment:
Any reason this is called updateVisitedState vs updateVisitedNode in, respectively, ShortestPathRecursiveJoin and VariableLengthJoin?